src/test/lro-poller.spec.ts

```diff
@@ -0,0 +1,126 @@
+import { expect } from "chai";
+import * as nock from "nock";
+
+import { LroPoller, LroPollerOptions } from "../lro-poller";
+import TimeoutError from "../throttler/errors/timeout-error";
+
+const TEST_ORIGIN = "https://firebasedummy.test.com";
+const VERSION = "v1";
+const LRO_RESOURCE_NAME = "operations/cp.3322442424242444";
+
+describe("LroPoller", () => {
+  describe("poll", () => {
+    const poller = new LroPoller();
+    let pollerOptions: LroPollerOptions;
+
+    beforeEach(() => {
+      pollerOptions = {
+        apiOrigin: TEST_ORIGIN,
+        apiVersion: VERSION,
+        operationResourceName: LRO_RESOURCE_NAME,
+        backoff: 10,
+      };
+    });
+
+    afterEach(() => {
+      nock.cleanAll();
+    });
+
+    it("should return result with response field if polling succeeds", async () => {
+      const successfulResponse = {
+        done: true,
+        response: "completed",
+      };
+      nock(TEST_ORIGIN)
+        .get(`/${VERSION}/${LRO_RESOURCE_NAME}`)
+        .reply(200, successfulResponse);
+
+      expect(await poller.poll(pollerOptions)).to.deep.equal({ response: "completed" });
+      expect(nock.isDone()).to.be.true;
+    });
+
+    it("should return result with error field if polling operation returns failure response", async () => {
+      const failedResponse = {
+        done: true,
+        error: {
+          message: "failed",
+          code: 7,
+        },
+      };
+      nock(TEST_ORIGIN)
+        .get(`/${VERSION}/${LRO_RESOURCE_NAME}`)
+        .reply(200, failedResponse);
+
+      const response = await poller.poll(pollerOptions);
+      expect(response.error.message).to.equal("failed");
+      expect(response.error.status).to.equal(7);
+      expect(nock.isDone()).to.be.true;
+    });
+
+    it("should return result with error field if api call rejects with unrecoverable error", async () => {
+      nock(TEST_ORIGIN)
+        .get(`/${VERSION}/${LRO_RESOURCE_NAME}`)
+        .reply(404, { context: { response: { statusCode: 404 } } });
```

Comment:
Can you test returning an error message here too?

Reply:
I tried, but the message was "HTTP 404: Not Found" (and "HTTP <CODE>: Unknown Response" for a lot of other codes, including 500 and 503 instead of the standard messages). I this probably decided by `api` util implementation.

I can do something like this though, but the assert is quite ugly:
 ```javascript
nock(TEST_ORIGIN)
        .get(`/${VERSION}/${LRO_RESOURCE_NAME}`)
        .reply(404, { context: { response: { statusCode: 404, message: "failed" } } });
const response = await poller.poll(pollerOptions);

expect(response.error.body.context.response.messge).to.equal("failed");
```

Note that the `body.context` contains much more data in it and I can just use `deep.equal`

Reply:
`response` should be of the type `LroResult`, right? So `response.error` should be of the type `FirebaseError`, which I don't think is supposed contain the field `body`. So if that field does exist here, that indicates to me that there's a bug that should be fixed.

What I'd expect to be working in the end is `expect(response.error.messge).to.equal("failed");`.

Reply:
Sorry for the confusion. `body` should be inside the `context` field of the `FirebaseError`. So I changed the test into:
```javascript
nock(TEST_ORIGIN)
  .get(`/${VERSION}/${LRO_RESOURCE_NAME}`)
  .reply(404, { message: "failed" });
const response = await poller.poll(pollerOptions);

expect(response.error.status).to.equal(404);
expect(response.error.context.body.message).to.equal("failed");
```
**Note:**  actual message of the error is determined [here](https://github.com/firebase/firebase-tools/blob/master/src/responseToError.js#L27)